### ISSUE TITLE
[Feature Request] Add restart_backend from pyasic as ha service

### ISSUE THREAD

**Issue by pixeldoc2000** [Schnitzel/hass-miner#361]
I use hass-miner mainly to control my Miner with Braiins Firmware (BOS) to make use of my excess Solar Energy, which works great so far. Thanks for the great ha integration.

My Problem is as follows:
I monitor the hashrate and pause and resume the miner if the hashrate drops below e certain level.

Sometime pause and resume does not restart the mining process, when the pool is "dead" according to BOS:
![grafik](https://github.com/Schnitzel/hass-miner/assets/376715/7cf00460-978e-4c84-9d2d-d0ae1b5f33e5)

To fix this situation, a restart of bosminer is required, but that is currently not possible via hass-miner. It would be nice to make `restart_backend()` from pyasic available as a ha service for example, to fix this problem (https://docs.pyasic.org/en/latest/miners/functions/#pyasic.miners.base.MinerProtocol.restart_backend).

Additionally, having the pool status available in HA would be awesome, but also cumbersome?

**Comment by b-rowan** [comment]
I like the restart backend idea, I'm not 100% sure how that would work regarding home assistants UI, but I assume it would be pretty simple.

Regarding the pool status, there is currently a Summer of Bitcoin project going on that is adding pool metrics to pyasic, so once that is done it will be a lot easier to show the active pool.  This is a bit more complicated because there can be multiple active pools (in the case of quota splitting between groups), but I think it could be displayed as `Active Pool(s): 1, 4`

**Comment by pixeldoc2000** [comment]
> I'm not 100% sure how that would work regarding home assistants UI, but I assume it would be pretty simple.
> 

My suggesten would be a service like `miner.restart_backend` , because HA UI still has no momentary button. The service could be called by automation or script (button).

**Comment by b-rowan** [comment]
This is essentially what I'm attempting, but sadly the dev docs for Hass are terrible.  I think I have something close to working, but it will have to take in an IP address, rather than allowing you to select a device.

**Comment by b-rowan** [comment]
Done!  Try out v1.1.10.

**Comment by pixeldoc2000** [comment]
Awesome, thx a lot! That was fast.

I will test it today and report here.

readme.md should to updated to include the new service calls.

**Comment by b-rowan** [comment]
Will do.  Gonna leave this open for the pools request.

**Comment by pixeldoc2000** [comment]
Hm, sadly it does not work as expected:
- `reboot` does trigger a ssh session to the miner ip, but does not reboot the miner. Executing `/sbin/reboot` manually via ssh does work.
- `restart_backend` does trigger a ssh session to the miner ip, but does not restart the bosminer. Executing`/etc/init.d/bosminer restart` manually via ssh does work.

I did not find a way to capture the commands send via ssh from HA / pyasic. All I can see is the incoming ssh connection with tcpdump on the miner from HA.

Tested with Braiins OS+ Version: 2022-09-27-0-26ba61b9-22.08.1-plus on Antminer S9.

**Comment by b-rowan** [comment]
That's really weird, I was able to get it working with my own S9, same firmware version and everything.  Is it possible the SSH connection is failing?  Did you change the SSH password or something?  Usually it should login with root no password...

**Comment by pixeldoc2000** [comment]
Yes, changed the root password.

**Comment by b-rowan** [comment]
Try removing and re-adding the miner.  It should prompt you for an SSH user and password, try changing it there...

**Comment by pixeldoc2000** [comment]
Readded the miner, entered ssh password in HA config flow. SSH connection still did not work.

```
# tail -f /var/log/syslog
Fri Jun 21 23:07:22 2024 authpriv.warn dropbear[3704]: Bad password attempt for 'root' from 192.168.0.5:56266
Fri Jun 21 23:07:22 2024 authpriv.info dropbear[3704]: Exit before auth from <192.168.0.5:56266>: (user 'root', 1 fails): Exited normally
```


Changed root to default password "root". Now it works.

Looks like the ssh password from the config flow is not used, instead pyasic probably falls back to the default password "root"?

**Comment by b-rowan** [comment]
Interesting.  That will almost certainly be an issue with pyasic, lemme do some testing.

**Comment by pixeldoc2000** [comment]
Yes, seems to be a bug in current `pyasic`. I have tested it with pyasic 0.57.6 again Braiins OS with non default password. With default password it works.

![grafik](https://github.com/Schnitzel/hass-miner/assets/376715/4dc1089f-5478-4833-9f3e-376e3ea9d024)

**Comment by pixeldoc2000** [comment]
BTW, you forgot to update `custom_components/miner/manifest.json` for 1.1.10 release.

**Comment by b-rowan** [comment]
Interestingly, the error you sent is not SSH related, it's web related.  If you comment out the web password setting, does that work?  It's possible the LUCI endpoints just use the default info, since technically they were removed a few versions before latest for the S9, but there is a special case where they return data manually.

**Comment by pixeldoc2000** [comment]
You are right, I forgot.

bosminer_restart is done via ssh. Data is gathered via LUCI API.

> Interestingly, the error you sent is not SSH related, it's web related. If you comment out the web password setting, does that work? It's possible the LUCI endpoints just use the default info, since technically they were removed a few versions before latest for the S9, but there is a special case where they return data manually.

I have only tested to get data via LUCI API, not ssh.

If the default password `root` is set on the miner firmware and `settings.update("default_bosminer_web_password", "changeme")` is commented out, LUCI API query works. If I change the password of user `root` on the miner to `changeme`, LUCI query fails with `APIError`.

**Comment by pixeldoc2000** [comment]
I am not able to create a quick test for `restart_backend()`, as I am not very deep into python.

```python
import asyncio
from pyasic import get_miner

#settings.update("default_bosminer_ssh_password", "changeme")

async def do_restart_backend():
    miner = await get_miner("192.168.0.203")

    # call control function
    result = await miner.restart_backend()

if __name__ == "__main__":
    asyncio.run(do_restart_backend())
```

# Error
```
Traceback (most recent call last):
  File "z:\xxx\braiinsos\pyasic-test-03.py", line 13, in <module>
    asyncio.run(do_restart_backend())
  File "C:\Users\xxx\AppData\Local\Programs\Python\Python311\Lib\asyncio\runners.py", line 190, in run
    return runner.run(main)
           ^^^^^^^^^^^^^^^^
  File "C:\Users\xxx\AppData\Local\Programs\Python\Python311\Lib\asyncio\runners.py", line 118, in run
    return self._loop.run_until_complete(task)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "C:\Users\xxx\AppData\Local\Programs\Python\Python311\Lib\asyncio\base_events.py", line 654, in run_until_complete
    return future.result()
           ^^^^^^^^^^^^^^^
  File "z:\xxx\braiinsos\pyasic-test-03.py", line 10, in do_restart_backend
    result = await miner.restart_backend()
```
Can you help me?

**Comment by b-rowan** [comment]
I see nothing wrong with that?  I think I understand what the issue is though.  First thing I should do is fix the LUCI password issues, I wasn't aware they did that...

**Comment by pixeldoc2000** [comment]
Looks like there are still some issue left with hass-miner 1.1.12

I have setup two miner in a test ha container:
![grafik](https://github.com/user-attachments/assets/b0de88fb-caf5-4842-9dea-e3d259b1775d)

Neither `miner: reboot` or `miner: restart_backend` service let me select any one miner either as a device or entitiy:
![grafik](https://github.com/user-attachments/assets/9b4d9919-0e09-4371-b8d7-2dd082fbc4c1)

Looks like the filter for the service does not work as expected or I am doing it wrong?

It did work with hass-miner 1.1.11, because one had to specify the ip address of the miner instead of selecting a device.

**Comment by b-rowan** [comment]
Ok, yeah. I'm not super familiar with the internals of HASS, so I can try to figure out how to do that.

**Comment by pixeldoc2000** [comment]
> Neither `miner: reboot` or `miner: restart_backend` service let me select any one miner either as a device or entitiy: 

My PR should fix the service call target filter.

The Error with the Service call itself remains. I get the following error when executing `miner: restart_backend` with a miner device selected:
```python
Traceback (most recent call last):
File "/usr/local/lib/python3.12/site-packages/homeassistant/components/websocket_api/decorators.py", line 28, in _handle_async_response
await func(hass, connection, msg)
File "/usr/local/lib/python3.12/site-packages/homeassistant/components/websocket_api/commands.py", line 793, in handle_execute_script
script_result = await script_obj.async_run(
^^^^^^^^^^^^^^^^^^^^^^^^^^^
File "/usr/local/lib/python3.12/site-packages/homeassistant/helpers/script.py", line 1770, in async_run
return await asyncio.shield(create_eager_task(run.async_run()))
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
File "/usr/local/lib/python3.12/site-packages/homeassistant/helpers/script.py", line 465, in async_run
await self._async_step(log_exceptions=False)
File "/usr/local/lib/python3.12/site-packages/homeassistant/helpers/script.py", line 529, in _async_step
self._handle_exception(
File "/usr/local/lib/python3.12/site-packages/homeassistant/helpers/script.py", line 559, in _handle_exception
raise exception
File "/usr/local/lib/python3.12/site-packages/homeassistant/helpers/script.py", line 527, in _async_step
await getattr(self, handler)()
File "/usr/local/lib/python3.12/site-packages/homeassistant/helpers/script.py", line 764, in _async_call_service_step
response_data = await self._async_run_long_action(
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
File "/usr/local/lib/python3.12/site-packages/homeassistant/helpers/script.py", line 727, in _async_run_long_action
return await long_task
^^^^^^^^^^^^^^^
File "/usr/local/lib/python3.12/site-packages/homeassistant/core.py", line 2731, in async_call
response_data = await coro
^^^^^^^^^^
File "/usr/local/lib/python3.12/site-packages/homeassistant/core.py", line 2774, in _execute_service
return await target(service_call)
^^^^^^^^^^^^^^^^^^^^^^^^^^
File "/config/custom_components/miner/services.py", line 36, in reboot
miner = await get_miner(call)
^^^^^^^^^^^^^^^^^^^^^
File "/config/custom_components/miner/services.py", line 28, in get_miner
if miner_id is None or miner_id not in miners:
^^^^^^^^^^^^^^^^^^^^^^
TypeError: unhashable type: 'list'
```

**Comment by b-rowan** [comment]
Should be fixed by implementing the changes from your PR (along with my fix for the above) in v1.1.14
